[JSC] Enable Allocation Sinking for NewArrayWithConstantSize

hyjorc1 opened the pull request:
#### 3353f1290c6aa2eebdd48c99c539a6a0858c8ab1
<pre>
[JSC] Enable Allocation Sinking for NewArrayWithConstantSize
<a href="https://bugs.webkit.org/show_bug.cgi?id=287731">https://bugs.webkit.org/show_bug.cgi?id=287731</a>
<a href="https://rdar.apple.com/144885784">rdar://144885784</a>

Reviewed by Yusuke Suzuki.

This patch enables allocation sinking for NewArrayWithConstantSize, allowing
the DFG JIT to eliminate unnecessary array allocations when safe. This
optimization removes dead allocations and materializes them only if needed,
reducing memory overhead and improving execution efficiency. See the comments
in DFGObjectAllocationSinkingPhase.cpp for details.

Changes:
1. Introduced PhantomNewArrayWithConstantSize and MaterializeNewArrayWithConstantSize
   nodes for sinking and materializing arrays.
2. Tracked array allocations and indexed properties using ArrayIndexedPropertyPLoc
   and ArrayLengthPropertyPLoc.
3. Eliminated redundant bounds checks with removeCheckInBoundsIfNeeded.
4. Implemented JIT and FTL support for materializing sunken arrays.
5. Guarded sinking with isWatchingArrayPrototypeChainIsSaneWatchpoint, isInBounds,
   and constant index access to ensure deoptimization safety.
6. Added Options::useArrayAllocationSinking for runtime control.

* JSTests/microbenchmarks/loop-unrolling-4.js:
(test):
* JSTests/microbenchmarks/loop-unrolling-5.js: Added.
(assert):
(test):
* JSTests/stress/array-allocation-sink.js: Added.
(assert):
(run):
(assert.test):
(run.test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::convertToPhantomNewArrayWithConstantSize):
(JSC::DFG::Node::hasNewArraySize):
(JSC::DFG::Node::newArraySize):
(JSC::DFG::Node::hasIndexingType):
(JSC::DFG::Node::hasObjectMaterializationData):
(JSC::DFG::Node::isPhantomAllocation):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGObjectMaterializationData.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGPromotedHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGPromotedHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGValidate.cpp:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLExitTimeObjectMaterialization.cpp:
(JSC::FTL::ExitTimeObjectMaterialization::ExitTimeObjectMaterialization):
* Source/JavaScriptCore/ftl/FTLExitTimeObjectMaterialization.h:
(JSC::FTL::ExitTimeObjectMaterialization::indexingType const):
(JSC::FTL::ExitTimeObjectMaterialization::size const):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileNewArrayWithConstantSizeImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileNewArrayWithConstantSize):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/IndexingType.h:
(JSC::isNewArrayWithConstantSizeIndexingType):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/290691@main">https://commits.webkit.org/290691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/488c182dc5d172b005d27a6b513c751c1e91ebc9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90712 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95727 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41498 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10643 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69795 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27341 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93713 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82257 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50137 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36642 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40627 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83544 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78195 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37706 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97557 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89492 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17907 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13152 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78086 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78030 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22465 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21088 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14312 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17916 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23259 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111981 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17655 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21111 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19439 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectMaterializationData.h:
```diff
@@ -38,7 +38,7 @@ namespace JSC { namespace DFG {
     // Determines the meaning of the passed nodes.
     Vector<PromotedLocationDescriptor> m_properties;
-    
+    unsigned m_newArraySize;
```
Let's make it zero initialized via `m_newArraySize { 0 };`

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -13200,6 +13200,47 @@ void SpeculativeJIT::compileLazyJSConstant(Node* node)
+        }
+        case ALL_INT32_INDEXING_TYPES:
+        case ALL_CONTIGUOUS_INDEXING_TYPES: {
+            JSValueOperand value(this, edge, ManualOperandSpeculation);
```
Do `speculate`

hyjorc1 replied:
Updated. Let's do speculate for all edges.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14298,22 +14339,24 @@ void SpeculativeJIT::compileNewArrayWithSize(Node* node)
+
+void SpeculativeJIT::compileNewArrayWithConstantSize(Node* node)
+{
+
```
Remove this empty line.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14298,22 +14339,24 @@ void SpeculativeJIT::compileNewArrayWithSize(Node* node)
+{
+
+    GPRTemporary size(this);
+    GPRTemporary result(this);
```
Since `gpr()` can have side effect, we would like to call them a-priori.

```

GPRReg sizeGPR = size.gpr();
GPRReg resultGPR = result.gpr();

```

and use `sizeGPR` and `resultGPR`.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -16268,6 +16277,37 @@ IGNORE_CLANG_WARNINGS_END
+                break;
+            case ALL_INT32_INDEXING_TYPES:
+            case ALL_CONTIGUOUS_INDEXING_TYPES: {
+                m_out.store64(lowJSValue(edge, ManualOperandSpeculation), butterfly, m_heaps.forIndexingType(indexingType)->at(index));
```
Need to call `speculate` onto `lowJSValue` results.

hyjorc1 replied:
Updated. Let's do speculate for all edges.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,119 @@ class ObjectAllocationSinkingPhase : public Phase {
+            } else {
+                m_graph.doToChildren(
+                    node,
+                    [&] (Edge edge) {
```
Remove space between `[&]` and `(Edge`

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,119 @@ class ObjectAllocationSinkingPhase : public Phase {
+            } else {
+                m_graph.doToChildren(
+                    node,
+                    [&] (Edge edge) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,119 @@ class ObjectAllocationSinkingPhase : public Phase {
+            //    incorrect behavior.
+            if (Options::useArrayAllocationSinking()) {
+                if (!m_graph.isWatchingArrayPrototypeChainIsSaneWatchpoint(node))
+                    break;
```
We need to escape in this case, right?

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,119 @@ class ObjectAllocationSinkingPhase : public Phase {
+                target = &m_heap.newAllocation(node, Allocation::Kind::Array);
+
+                if (!m_sinkCandidates.contains(node) || !enableLengthMaterialization)
+                    break;
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,119 @@ class ObjectAllocationSinkingPhase : public Phase {
+            target = m_heap.onlyLocalAllocation(base);
+            if (target && target->isArrayAllocation()) {
+                if (node->op() == GetArrayLength && m_sinkCandidates.contains(base))
+                    exactRead = PromotedLocationDescriptor(ArrayLengthPropertyPLoc);
```
Can you explain (and leaving comment about) how `GetButterfly` case is handled?
Also, don't we need to do escape when `m_sinkCandidates.contains(base)` is false?

hyjorc1 replied:
We've already ensured that once `GetByVal` and `PutByVal` are valid then all the related `CheckInBounds` and `GetArrayLength` in the dependency graph will be gone. 

Since `Got ByVal`, `Put ByVal`, and `GetArrayLength` are the only referrers, once they are eliminated then the corresponding `GetButterfly` must be removed in the late cleanup phase.

Maybe we should just remove  `GetButterfly` explicitly in this phase?

hyjorc1 replied:
Just like how we remove the `CheckInBounds`.

hyjorc1 replied:
Updated with explicit removal.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1285,6 +1408,28 @@ class ObjectAllocationSinkingPhase : public Phase {
+            for (unsigned i = 0; i < node->numChildren(); ++i) {
+                Node* child = m_graph.varArgChild(node, i).node();
+                if (child->op() == CheckInBounds) {
+                    child->decRef();
```
I wonder if this ref count is still intact or not in this phase.

hyjorc1 replied:
For `NewArrayWithConstantSize` related nodes in that dependency graph, I think so. Conservatively, we can recompute the ref count just before  `handleNode` with `enableLengthMaterialization = true`.

hyjorc1 replied:
Updated with explicit `m_graph.computeRefCounts()`.

hyjorc1 replied:
Removed `removeArrayDependentNodesIfNeeded `.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1266,6 +1376,7 @@ class ObjectAllocationSinkingPhase : public Phase {
             ASSERT(writes.isEmpty());
             if (Node* value = heapResolve(PromotedHeapLocation(target->identifier(), exactRead))) {
                 ASSERT(!value->replacement());
+                removeArrayDependentNodesIfNeeded(node);
```
Is it guaranteed to be safe? Isn't it possible that GetByVal / PutByVal's index is oob?

hyjorc1 replied:
I think so. All lowerable `GetByVal` and `PutByVal` are guarded by `arrayMode.isInBounds()` see the `GetByVal` and `PutByVal` cases in `hanldeNode`.

hyjorc1 replied:
Removed `removeArrayDependentNodesIfNeeded `.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,106 @@ class ObjectAllocationSinkingPhase : public Phase {
+                    Edge value = m_graph.varArgChild(node, 2);
+                    writes.add(PromotedLocationDescriptor(ArrayIndexedPropertyPLoc, index->asInt32()), LazyNode(value.node()));
+                    if (m_sinkCandidates.contains(base))
+                        m_nodeToUseKind.add(value.node(), value.useKind());
```
I think OSR exit via this check in materialization code sounds not correct. Rather, we should emit Check with this UseKind here (so, I don't think we need  m_nodeToUseKind)

hyjorc1 replied:
The reason to update UseKind for the edges of `MaterializeNewArrayWithConstantSize` is not only for the speculate check in the compile phase. But also for passing the validation checks for double edges of `MaterializeNewArrayWithConstantSize` when the indexing type is in the double cases. Any suggestions, should we just insert `Check` nodes here and by pass the double edge validations?

hyjorc1 replied:
Updated with an inserted `Check`.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2058,6 +2207,8 @@ class ObjectAllocationSinkingPhase : public Phase {
                 bool desiredNextExitOK = node->origin.exitOK && !clobbersExitState(m_graph, node);
 
                 bool doLower = false;
+                m_nodeIndex = nodeIndex;
```
Let's not have this member variable. `handleNode` is used in multiple places but only this place is setting m_nodeIndex, that sounds dangerous. We should pass nodeIndex to handleNode if it is necessary.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2058,6 +2207,8 @@ class ObjectAllocationSinkingPhase : public Phase {
 
                 bool doLower = false;
+                m_nodeIndex = nodeIndex;
+                bool enableLengthMaterialization = true;
```
Let's remove this as it is too complicated.

hyjorc1 replied:
We only need to materialize the constant (the replacment) for `GetArrayLength` once which is at this site. And `handleNode` is used in multiple places.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -935,6 +944,106 @@ class ObjectAllocationSinkingPhase : public Phase {
+                if (!m_sinkCandidates.contains(node) || !enableLengthMaterialization)
+                    goto escapeChildren;
+                Node* length = m_insertionSet.insertNode(m_nodeIndex, SpecInt32Only, JSConstant, node->origin,
+                    OpInfo(m_graph.freeze(jsNumber(node->newArraySize()))));
```
We should just add it to the root block and use it. We can track whether we need to insert a new constant via `UncheckedKeyHashMap<unsigned, Node*, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>>`

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -921,6 +943,13 @@ class ObjectAllocationSinkingPhase : public Phase {
+    Node* ensureConstant(unsigned constant)
+    {
+        return m_constants.ensure(constant, [&]() {
+            return m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(constant));
```
This is not correct since `m_insertionSet` needs to be executed with root block, but it isn't. We should have different InsertionSet `m_rootInsertionSet`, and we need to run `m_rootInsertionSet.execute(rootBlock)` later.

hyjorc1 committed:
[JSC] Enable Allocation Sinking for NewArrayWithConstantSize
https://bugs.webkit.org/show_bug.cgi?id=287731
rdar://144885784

Reviewed by Yusuke Suzuki.

This patch enables allocation sinking for NewArrayWithConstantSize, allowing
the DFG JIT to eliminate unnecessary array allocations when safe. This
optimization removes dead allocations and materializes them only if needed,
reducing memory overhead and improving execution efficiency. See the comments
in DFGObjectAllocationSinkingPhase.cpp for details.

Changes:
1. Introduced PhantomNewArrayWithConstantSize and MaterializeNewArrayWithConstantSize
   nodes for sinking and materializing arrays.
2. Tracked array allocations and indexed properties using ArrayIndexedPropertyPLoc
   and ArrayLengthPropertyPLoc.
3. Eliminated redundant bounds checks with removeCheckInBoundsIfNeeded.
4. Implemented JIT and FTL support for materializing sunken arrays.
5. Guarded sinking with isWatchingArrayPrototypeChainIsSaneWatchpoint, isInBounds,
   and constant index access to ensure deoptimization safety.
6. Added Options::useArrayAllocationSinking for runtime control.

* JSTests/microbenchmarks/loop-unrolling-4.js:
(test):
* JSTests/microbenchmarks/loop-unrolling-5.js: Added.
(assert):
(test):
* JSTests/stress/array-allocation-sink.js: Added.
(assert):
(run):
(assert.test):
(run.test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::convertToPhantomNewArrayWithConstantSize):
(JSC::DFG::Node::hasNewArraySize):
(JSC::DFG::Node::newArraySize):
(JSC::DFG::Node::hasIndexingType):
(JSC::DFG::Node::hasObjectMaterializationData):
(JSC::DFG::Node::isPhantomAllocation):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGObjectMaterializationData.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGPromotedHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGPromotedHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGValidate.cpp:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLExitTimeObjectMaterialization.cpp:
(JSC::FTL::ExitTimeObjectMaterialization::ExitTimeObjectMaterialization):
* Source/JavaScriptCore/ftl/FTLExitTimeObjectMaterialization.h:
(JSC::FTL::ExitTimeObjectMaterialization::indexingType const):
(JSC::FTL::ExitTimeObjectMaterialization::size const):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileNewArrayWithConstantSizeImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileNewArrayWithConstantSize):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/IndexingType.h:
(JSC::isNewArrayWithConstantSizeIndexingType):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/290691@main
